core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java

```diff
@@ -19,201 +19,103 @@
 
 import java.io.IOException;
-import java.util.*;
 
 public class QueryParseContext {
 
     private static final ParseField CACHE = new ParseField("_cache").withAllDeprecated("Elasticsearch makes its own caching decisions");
     private static final ParseField CACHE_KEY = new ParseField("_cache_key").withAllDeprecated("Filters are always used as cache keys");
 
-    private static ThreadLocal<String[]> typesContext = new ThreadLocal<>();
-
-    public static void setTypes(String[] types) {
-        typesContext.set(types);
-    }
-
-    public static String[] getTypes() {
-        return typesContext.get();
-    }
-
-    public static String[] setTypesWithPrevious(String[] types) {
-        String[] old = typesContext.get();
-        setTypes(types);
-        return old;
-    }
-
-    public static void removeTypes() {
-        typesContext.remove();
-    }
-
-    private final Index index;
-
-    private final Version indexVersionCreated;
-
-    private final IndexQueryParserService indexQueryParser;
-
-    private final Map<String, Query> namedQueries = Maps.newHashMap();
-
-    private final MapperQueryParser queryParser = new MapperQueryParser(this);
-
     private XContentParser parser;
-
+    private final Index index;
+    private boolean isFilter;
     private ParseFieldMatcher parseFieldMatcher;
 
-    private boolean allowUnmappedFields;
-
-    private boolean mapUnmappedFieldAsString;
-
-    private NestedScope nestedScope;
-
-    private boolean isFilter;
+    private final QueryCreationContext context;
+    private IndicesQueriesRegistry indicesQueriesRegistry;
 
     public QueryParseContext(Index index, IndexQueryParserService indexQueryParser) {
-        this.index = index;
-        this.indexVersionCreated = Version.indexCreated(indexQueryParser.indexSettings());
-        this.indexQueryParser = indexQueryParser;
-    }
-
-    public void parseFieldMatcher(ParseFieldMatcher parseFieldMatcher) {
-        this.parseFieldMatcher = parseFieldMatcher;
+        this(new QueryCreationContext(index, indexQueryParser));
     }
 
-    public ParseFieldMatcher parseFieldMatcher() {
-        return parseFieldMatcher;
+    public QueryParseContext(QueryCreationContext context) {
+        this.context = context;
+        this.index = context.index();
+        this.indicesQueriesRegistry = context.indexQueryParserService().indicesQueriesRegistry();
     }
 
     public void reset(XContentParser jp) {
-        allowUnmappedFields = indexQueryParser.defaultAllowUnmappedFields();
         this.parseFieldMatcher = ParseFieldMatcher.EMPTY;
-        this.lookup = null;
         this.parser = jp;
-        this.namedQueries.clear();
-        this.nestedScope = new NestedScope();
-        this.isFilter = false;
+        this.context.reset();
     }
 
-    public Index index() {
-        return this.index;
-    }
-
-    public void parser(XContentParser parser) {
-        this.parser = parser;
+    public QueryCreationContext context() {
```

Comment:
Also, will this stay around or go away at some point? If it's supposed to be gone we should mark as //norelease